### PR TITLE
[mempool] Wait until max block size or quorum_store_poll_count is reached

### PR DESCRIPTION
**Pull request by igor-aptos**
Adding two new flags, to improve blockchain performance.

   // Whether to create partial blocks when few transactions exist, or empty blocks when there is
    // pending ordering, or to wait for quorum_store_poll_count * 30ms to collect transactions for a block
    //
    // It is more efficient to execute larger blocks, as it creates less overhead. On the other hand
    // waiting increases latency (unless we are under high load that added waiting latency
    // is compensated by faster execution time). So we want to balance the two, by waiting only
    // when we are saturating the execution pipeline:
    // - if there are more pending blocks then usual in the execution pipeline,
    //   block is going to wait there anyways, so we can wait to create a bigger/more efificent block
    // - in case our node is faster than others, and we don't have many pending blocks,
    //   but we still see very large recent (pending) blocks, we know that there is demand
    //   and others are creating large blocks, so we can wait as well.

### CODE REVIEW

**Comment by gelash** [comment]
Should we instead have an optional minimum threshold? Seems a bit more general, setting to the upper limit should be equivalent to disabling non-full boolean

**Comment by igor-aptos** [comment]
Now that quorum store has landed, rebased on main, and re-sending this for review. 

we can see if we want to change defaults immediatelly, or later. Also whether we want to have choice here depends on recent history of block sizes (i.e. wait for full blocks, if last 10 blocks averaged more than 100 txns/block)

**Comment by igor-aptos** [comment]
I would like to land this - we can make default configs not change behavior for now?

@zekun000 , @bchocho , @sitalkedia  - can I get a review?

File: consensus/src/quorum_store/utils.rs [hunk 1]
```diff
@@ -314,13 +319,30 @@ impl ProofQueue {
                         .observe((current_time.round() - expiration.round()) as f64);
                 }
             }
+            size = size - 1;
         }
         counters::EXPIRED_PROOFS_WHEN_PULL.observe(num_expired_but_not_committed as f64);
         counters::BLOCK_SIZE_WHEN_PULL.observe(cur_txns as f64);
         counters::BLOCK_BYTES_WHEN_PULL.observe(cur_bytes as f64);
         counters::PROOF_SIZE_WHEN_PULL.observe(ret.len() as f64);
```
Comment (ibalajiarun) [inline consensus/src/quorum_store/utils.rs:327]: Should we move this into the if block below, so we only count when we return a "good" result?

**Comment by sitalkedia** [comment]
@igor-aptos - On a high level, what do we think about making this more adaptive/smarter? For example, the wait time can be dynamic based on the back-pressure value (number of outstanding blocks in the pipeline) - if there are several blocks in the pipeline waiting to be executed/committed, we can wait longer. If not, then the wait can be shorter.  We can periodically poll the number of outstanding blocks and decide to create a block when we either reach the max block size limit or the number of pending blocks < some threshold. Thoughts?

File: consensus/src/quorum_store/utils.rs (hunk 1 above)
Comment (igor-aptos) [reply consensus/src/quorum_store/utils.rs]: fixed

**Comment by igor-aptos** [comment]
@sitalkedia - made it adaptive, based on the size of largest pending block

**Comment by sitalkedia** [comment]
Overall looks reasonable to me - left minor comments. Can you also address my previous comments?

File: config/src/config/consensus_config.rs [hunk 2]
```diff
@@ -36,9 +36,22 @@ pub struct ConsensusConfig {
...
+    // It is more efficient to execute larger blocks, as it creates less overhead. On the other hand
+    // waiting increases latency (unless we are under high load that added waiting latency
+    // is compensated by faster execution time). So we want to balance the two, by waiting only
+    // when we are saturating the execution pipeline:
+    // - if there are more pending blocks then usual in the execution pipeline,
+    //   block is going to wait there anyways, so we can wait to create a bigger/more efificent block
+    // - in case our node is faster than others, and we don't have many pending blocks,
+    //   but we still see very large recent (pending) blocks, we know that there is demand
+    //   and others are creating large blocks, so we can wait as well.
+    pub wait_for_full_blocks_above_num_pending_uncommitted_blocks: usize,
```
Comment (sitalkedia) [inline config/src/config/consensus_config.rs:51]: How about the names? Can we shorten it? May be `wait_until_pending_blocks` and `wait_above_fill_threshold` ?

File: consensus/src/epoch_manager.rs [hunk 3]
```diff
@@ -657,6 +657,8 @@ impl EpochManager {
             consensus_to_quorum_store_tx,
             self.config.quorum_store_poll_count, // TODO: consider moving it to a quorum store config in later PRs.
             self.config.quorum_store_pull_timeout_ms,
+            self.config.wait_for_full_blocks_above_recent_fill_threshold,
+            self.config.wait_for_full_blocks_above_num_pending_uncommitted_blocks,
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply consensus/src/epoch_manager.rs]: I mean sure, but then you have precision issues, etc. 

So I think it is better/simpler to have them separate - 10 and 8 currently

File: config/src/config/consensus_config.rs (hunk 2 above)
Comment (igor-aptos) [reply config/src/config/consensus_config.rs:51]: I prefer names that explain what they are about. especially for configurations, that can have undesirable effects if missunderstood.

wait_until_pending_blocks / wait_above_fill_threshold doesn't say what wait it is referring to. also pending blocks is not clear "pending" how, but I can shorten it to 

wait_for_full_blocks_above_pending_blocks

and keep the other the same?

File: consensus/src/payload_client.rs [hunk 4]
```diff
@@ -89,7 +97,12 @@ impl PayloadClient for QuorumStoreClient {
         exclude_payloads: PayloadFilter,
         wait_callback: BoxFuture<'static, ()>,
         pending_ordering: bool,
+        pending_uncommitted_blocks: usize,
+        recent_fill_fraction: f32,
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply consensus/src/payload_client.rs:101]: will use recent_max_fill_fraction

File: consensus/src/epoch_manager.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sitalkedia) [reply consensus/src/epoch_manager.rs]: More than precision, let's say we reduce the `back_pressure` value to 5, we would have to tune this config again. So it is probably better to have this as some fraction of the back_pressure and we can have that fraction configurable if needed?

[the remark this answers is not in the document]
Comment (igor-aptos) [reply consensus/src/epoch_manager.rs]: irrespective of the back pressure, normal value for pending blocks is around 4-6, so reducing it automatically below 6 will have negative consequences. 

testing smaller backpressure requires testing this property correctly as well, unfortunately. 

so I would keep it as is.

**Comment by sitalkedia** [comment]
Can you confirm above behavior change is only in case of Quorum store enabled and its a no-op without that?

**Comment by igor-aptos** [comment]
@sitalkedia - this change applies to everyone. (I wrote this PR before quorum store :) )

do you think we should disable it on main? if so, I can land disabled, and add config change to preview to enable it.

but I think it is useful to have it everywhere?

**Comment by sitalkedia** [comment]
@igor-aptos - How about we disable this in main and we can enable this after testing in previewnet?

**Comment by igor-aptos** [comment]
@sitalkedia done

**Review by sitalkedia — APPROVED** [review]

File: mempool/src/core_mempool/mempool.rs [hunk 5]
```diff
@@ -233,14 +236,21 @@ impl Mempool {
...
             LogSchema::new(LogEntry::GetBlock),
             seen_consensus = seen_size,
             walked = txn_walked,
             seen_after = seen.len(),
+            // before size and non full check
             result_size = result_size,
-            block_size = block.len(),
+            // before non full check
             byte_size = total_bytes,
+            block_size = block.len(),
```
Comment (zekun000) [inline mempool/src/core_mempool/mempool.rs:252]: this is after the clear, is this expected?

**Review by zekun000 — APPROVED** [review]

File: consensus/src/payload_client.rs [hunk 6]
```diff
@@ -110,9 +132,16 @@ impl PayloadClient for QuorumStoreClient {
             }
             break payload;
         };
-        debug!(
+        info!(
```
[the remark this answers is not in the document]
Comment (igor-aptos) [reply consensus/src/payload_client.rs:135]: proposal is not often - i.e. every tens of seconds. I think it is useful to have this on nodes not logging debug level.

**Review by ibalajiarun — COMMENTED** [review]
I have a high-level question. Why wait to create blocks? Can we not try to create either full blocks or empty blocks at all times and sometimes create partially full blocks if few transactions exist? I feel like there is some complexity to deciding if we need to propose or wait. There could be more edge cases no? What if a proposer realizes very late that they are the proposer but instead waits to form a full block. There is a higher chance the proposer is timed out? This may be one example that could be covered, but there could be others too right?

Instead, it is much easier to decide whether to populate a block or not. That way consensus keeps making progress regardless. wdyt?

**Comment by igor-aptos** [comment]
- there is a cost to empty blocks too (they are not empty, they just have no user transactions, BlockMetadata transaction still needs to be there)
- there is a waiting expectation already, so timeout values account for those. If we didn't have it, we could reduce the wait time though.
- more precise meaning here is - we wait 400ms, but end the wait early if we can create a full block
- once efficiency is such that there is no efficiency benefit from large blocks, we can remove this.

**Pull request merged**